### PR TITLE
Fix unauthorized support

### LINKED ISSUE #9758: Cannot connect to a remote kernel
Issue Type: <b>Bug</b>

On trying to connect to a remote kernel, the extension says something like this:
```
Error 08:31:54: Failed to find & set preferred controllers [i [Error]: https://<an ip address>:8888/
	at /home/mike/.vscode-server/extensions/ms-toolsai.jupyter-2022.3.1000901801/out/extension.js:2:368403
	at processTicksAndRejections (node:internal/process/task_queues:96:5)] {
  category: 'jupyterselfcert'
}
```

Extension version: 2022.3.1000901801
VS Code version: Code 1.66.2 (dfd34e8260c270da74b5c2d86d61aee4b6d56977, 2022-04-11T07:49:20.994Z)
OS version: Darwin x64 21.3.0
Restricted Mode: No
Remote OS version: Linux x64 5.4.0-96-generic

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M1 (8 x 24)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|2, 2, 2|
|Memory (System)|16.00GB (0.14GB free)|
|Process Argv|--crash-reporter-id de7537ee-b7ff-4aa6-9ac1-40d3f0a20bb2|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|SSH: seraph|
|OS|Linux x64 5.4.0-96-generic|
|CPUs|Intel Xeon Processor (Icelake) (14 x 1999)|
|Memory (System)|13.66GB (1.09GB free)|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyl392:30443607
pythontb:30283811
pythonvspyt551cf:30345471
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscorecescf:30445987
pythondataviewer:30285071
vscod805:30301674
pythonvspyt200:30340761
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
pythonvsnew555:30457759
vscscmwlcmt:30465135
cppdebugcf:30475217

```

</details>

### PR DESCRIPTION
**Pull request by rchiodo**
Fixes #9758 

Changed how we ask for 'allow unathorized'. This used to happen only when running a kernel. Kernels are now found just in the kernel finder so there was no ask at that point.

This code will now ask whenever we connect.

### CODE REVIEW

File: src/kernels/jupyter/launcher/jupyterExecution.ts [hunk 1]
```diff
@@ -190,6 +193,32 @@ export class JupyterExecutionBase implements IJupyterExecution {
         return Promise.resolve(undefined);
     }
 
+    public async validateRemoteUri(uri: string): Promise<void> {
```
Comment (rchiodo) [inline src/kernels/jupyter/launcher/jupyterExecution.ts:196]: Added a function to validate the URI. Might use this in the kernel finder too for cached items.

File: src/platform/errors/errorHandler.node.ts [hunk 2]
```diff
@@ -66,24 +67,7 @@ export class DataScienceErrorHandler implements IDataScienceErrorHandler {
...
-                            'allowUnauthorizedRemoteConnection',
-                            true,
-                            undefined,
-                            ConfigurationTarget.Workspace
-                        );
-                    } else if (value === closeOption) {
-                        sendTelemetryEvent(Telemetry.SelfCertsMessageClose);
-                    }
-                });
+            void handleCertsError(this.applicationShell, this.configuration, err.message);
```
Comment (rchiodo) [inline src/platform/errors/errorHandler.node.ts:70]: This should probably be an await. It was before.

File: src/kernels/jupyter/launcher/jupyterExecution.ts [hunk 3]
```diff
@@ -190,6 +193,32 @@ export class JupyterExecutionBase implements IJupyterExecution {
...
+        try {
+            // Prepare our map of server URIs (needed in order to retrieve the uri during the connection)
+            await this.updateServerUri(uri);
+
+            // Create an active connection.
+            connection = await createRemoteConnectionInfo(uri, this.getServerUri.bind(this));
+
+            // Attempt to list the running kernels. It will return empty if there are none, but will
+            // throw if can't connect.
+            sessionManager = await this.jupyterSessionManagerFactory.create(connection, false);
```
Comment (rchiodo) [inline src/kernels/jupyter/launcher/jupyterExecution.ts:208]: Password connect stuff should happen here too. This means we'll ask for user name password right after the user enters their URI

File: src/kernels/installer/moduleInstaller.node.ts [hunk 4]
```diff
@@ -184,27 +185,3 @@ export abstract class ModuleInstaller implements IModuleInstaller {
         flags?: ModuleInstallFlags
     ): Promise<ExecutionInstallArgs>;
 }
-
```
Comment (rchiodo) [inline src/kernels/installer/moduleInstaller.node.ts:187]: Move this to a non node specific file so errorHandler didn't have to be node specific anymore.

File: src/kernels/jupyter/launcher/jupyterExecution.ts (hunk 3 above)
Comment (DonJayamanne) [reply src/kernels/jupyter/launcher/jupyterExecution.ts:208]: perfect. was planning on adding that.

**Review by IanMatthewHuff — APPROVED** [review]

**Pull request merged**